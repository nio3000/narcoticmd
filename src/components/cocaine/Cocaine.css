.Cocaine {
  background: #AAA;
  padding: 20px;
  position: absolute;
  z-index: 10
}