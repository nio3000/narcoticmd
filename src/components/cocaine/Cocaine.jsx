import React, { Component } from 'react';
import { RadioGroup, Radio } from 'react-radio-group';
import './Cocaine.css';


class Cocaine extends Component {

  constructor(props) {
    super(props);
    this.state = {
      selectedValue: null,
      choosenPortion: null,
      key: ''
    };

    this.chooseFormatting = this.chooseFormatting.bind(this);
    this.handleShortcut = this.handleShortcut.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  /**
   * Choose formatting option
   * @param value ???
   */
  chooseFormatting(value) {
    this.setState({ selectedValue: value });
  }

  handleShortcut(event) {
    let key = event.target.value.toLocaleLowerCase();
    //Choose group
    let choosenPortion = this.props.portion.find((item) => item.key === key);
    //Prevent from reaction for undefined key
    if ( choosenPortion && !this.state.choosenPortion ) {
      this.setState({
        choosenPortion: choosenPortion
      });
      //TODO test
      //Return format if there is only one option (e.g. blockquote)
      if(choosenPortion.options.length === 1) {
        console.log(choosenPortion);
        this.props.onChooseFormat( choosenPortion.options[0].format );
        this.setState({
          key
        })
      }

    }
    else if ( this.state.choosenPortion ) {
      // Choose formatting option
      let option = this.state.choosenPortion.options.find((item) => item.key === key);
      if( !option ) {
        this.props.onChooseFormat(null);
        return;
      }
      this.props.onChooseFormat( option.format );
      this.setState({
        key
      })
    }
  }

  /**
   * Replace first occurence of shortcut in formatting option name
   * @param option string
   */
  displayShortcut( option ) {
    //
  }

  /**
   * Change Cocaine visible state in parent component
   */
  onBlur() {
    this.props.onBlur();
  }

  render() {
    let options = <label className="cocaine__hidden-option"><Radio/></label>;
    if(this.state.choosenPortion && this.state.choosenPortion.options) {
      options = this.state.choosenPortion.options.map(item =>
        <label
          key={item.key}
          className="cocaine__label"
        >
          <Radio value={item.key}/>{item.name}
        </label>
      );
    }
    return (
      <div className="cocaine" onBlur={this.onBlur}>
        <form>
          <input
            type="text"
            onChange={this.handleShortcut}
            onKeyDown={this.resetKey}
            className="cocaine__inhaler mousetrap"
            value={this.state.key}
            autoFocus
          />

          {this.props.portion.map( item =>
            <section className="cocaine__section" key={item.key}>
              <p className="cocaine__section-header">{ item.name }</p>
            </section>
          )}
          <section className="cocaine__section">
          <RadioGroup
            name="headers"
            selectedValue={this.setState.selectedValue}
            onChange={this.chooseFormatting}
            className="cocaine__section-options">
            {options}
          </RadioGroup>
          </section>
        </form>
      </div>
    );
  }
}


export default Cocaine;