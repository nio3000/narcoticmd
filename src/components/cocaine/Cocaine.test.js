import React from 'react';
import Cocaine from './Cocaine';
import renderer from 'react-test-renderer';

const portion = [
  {
    name: 'header',
    key: 'h',
    options: [
      { key: '1', name: 'h1', format: 'h1'},
      { key: '2', name: 'h2', format: 'h2'},
      { key: '3', name: 'h3', format: 'h3'},
    ]
  },
  {
    name: 'emphasis',
    key: 'e',
    options: [
      {key: 'b', name: 'b', format: 'bold'},
      {key: 'i', name: 'i', format: 'italic'},
      {key: 's', name: 's', format: 'strike'}
    ]
  },
  {
    name: 'blockquotes',
    key: 'q',
    options: [
      { name: 'quotation', key: 'q', format: '> {STR}' }
    ]
  },
  {
    name: 'list',
    key: 'l',
    options: [
      { name: 'ordered', key: 'o', format: '1. {STR}' },
      { name: 'unordered', key: 'u', format: '* {STR}' },
    ]
  }
];

describe('Cocaine', () => {
  test('renders correctly', () => {
   const component = renderer.create(
     <Cocaine portion={portion} />
   );
   let tree = component.toJSON();
   expect(tree).toMatchSnapshot();
  })
});

//Hide on blur

//Allow to delete selected text (x key in cocaine?)