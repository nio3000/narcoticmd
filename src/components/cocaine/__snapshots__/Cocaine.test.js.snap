// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cocaine renders correctly 1`] = `
<div
  className="cocaine"
  onBlur={[Function]}
>
  <form>
    <input
      autoFocus={true}
      className="cocaine__inhaler mousetrap"
      onChange={[Function]}
      onKeyDown={undefined}
      type="text"
      value=""
    />
    <section
      className="cocaine__section"
    >
      <p
        className="cocaine__section-header"
      >
        header
      </p>
    </section>
    <section
      className="cocaine__section"
    >
      <p
        className="cocaine__section-header"
      >
        emphasis
      </p>
    </section>
    <section
      className="cocaine__section"
    >
      <p
        className="cocaine__section-header"
      >
        blockquotes
      </p>
    </section>
    <section
      className="cocaine__section"
    >
      <p
        className="cocaine__section-header"
      >
        list
      </p>
    </section>
    <section
      className="cocaine__section"
    >
      <div
        className="cocaine__section-options"
      >
        <label
          className="cocaine__hidden-option"
        >
          <input
            name="headers"
            onChange={[Function]}
            type="radio"
          />
        </label>
      </div>
    </section>
  </form>
</div>
`;
